File: app/src/main/java/com/iita/akilimo/adapters/AdapterListAnimation.java
package com.iita.akilimo.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.iita.akilimo.R;
import com.iita.akilimo.models.Recommendations;
import com.iita.akilimo.utils.ItemAnimation;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class AdapterListAnimation extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Recommendations> items;

    private Context context;
    private OnItemClickListener mOnItemClickListener;
    private int animation_type;
    private int lastPosition = -1;
    private int layoutId;
    private boolean on_attach = true;

    public interface OnItemClickListener {
        void onItemClick(View view, Recommendations obj, int position);
    }

    public void setOnItemClickListener(final OnItemClickListener mItemClickListener) {
        this.mOnItemClickListener = mItemClickListener;
    }

    public AdapterListAnimation(Context context) {
        this.context = context;
        this.layoutId = R.layout.item_card_recommendation_arrow;
    }

    public AdapterListAnimation(Context context, int layoutId) {
        this.context = context;
        this.layoutId = layoutId;
    }

    public void setItems(List<Recommendations> items, int animation_type) {
        this.items = items;
        this.animation_type = animation_type;
        notifyDataSetChanged();
    }

    public class OriginalViewHolder extends RecyclerView.ViewHolder {
        public ImageView image;
        public TextView name;
        public View cardView;
        public View contentLayout;

        public OriginalViewHolder(View view) {
            super(view);
            name = view.findViewById(R.id.name);
            cardView = view.findViewById(R.id.lyt_parent);
            contentLayout = view.findViewById(R.id.contentLayout);
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;
        View view = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
        viewHolder = new OriginalViewHolder(view);
        return viewHolder;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NotNull RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof OriginalViewHolder) {
            OriginalViewHolder view = (OriginalViewHolder) holder;
            Recommendations rec = items.get(position);
            view.name.setText(rec.getRecommendationName());

            view.cardView.setOnClickListener(view1 -> {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(view1, items.get(position), position);
                }
            });
            setAnimation(view.itemView, position);
        }
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                on_attach = false;
                super.onScrollStateChanged(recyclerView, newState);
            }
        });
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }


    private void setAnimation(View view, int position) {
        if (position > lastPosition) {
            ItemAnimation.animate(view, on_attach ? position : -1, animation_type);
            lastPosition = position;
        }
    }

}

File: app/src/main/java/com/akilimo/mobile/adapters/AdapterListAnimation.kt
package com.akilimo.mobile.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.akilimo.mobile.databinding.ItemCardRecommendationArrowBinding
import com.akilimo.mobile.models.Recommendation
import com.akilimo.mobile.utils.TheItemAnimation.animate

@Deprecated(
    message = "Consider moving to a RecOptionsAdapter",
    replaceWith = ReplaceWith("RecOptionsAdapter")
)
class AdapterListAnimation :
    ListAdapter<Recommendation, AdapterListAnimation.OriginalViewHolder>(RecommendationDiffCallback()) {

    private var lastPosition = -1
    private var onAttach = true
    private var animationType: Int = 0
    private var onItemClickListener: ((View, Recommendation, Int) -> Unit)? = null

    fun setAnimationType(type: Int) {
        this.animationType = type
    }

    inner class OriginalViewHolder(val binding: ItemCardRecommendationArrowBinding) :
        RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): OriginalViewHolder {
        val binding = ItemCardRecommendationArrowBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return OriginalViewHolder(binding)
    }

    override fun onBindViewHolder(holder: OriginalViewHolder, position: Int) {
        val recommendation = getItem(position)
        with(holder.binding) {
            recTitle.text = recommendation.recommendationName

            recCard.setOnClickListener { view ->
                onItemClickListener?.invoke(view, recommendation, position)
            }
        }
        setAnimation(holder.itemView, position)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                onAttach = false
                super.onScrollStateChanged(recyclerView, newState)
            }
        })
    }

    private fun setAnimation(view: View, position: Int) {
        if (position > lastPosition) {
            animate(view, if (onAttach) position else -1, animationType)
            lastPosition = position
        }
    }


    fun setOnItemClickListener(listener: (View, Recommendation, Int) -> Unit) {
        this.onItemClickListener = listener
    }

    class RecommendationDiffCallback : DiffUtil.ItemCallback<Recommendation>() {
        override fun areItemsTheSame(oldItem: Recommendation, newItem: Recommendation): Boolean {
            return oldItem.recommendationName == newItem.recommendationName
        }

        override fun areContentsTheSame(
            oldItem: Recommendation,
            newItem: Recommendation
        ): Boolean {
            return oldItem.recommendationName == newItem.recommendationName
        }
    }
}

